gui/journal.lua

```diff
@@ -48,69 +150,174 @@ function JournalWindow:sanitizeFrame(frame)
     return frame
 end
 
-function JournalWindow:postUpdateLayout()
-    self:saveConfig()
-end
-
 function JournalWindow:saveConfig()
+    if not self.save_layout then
+        return
+    end
+
+    local toc = self.subviews.table_of_contents_panel
+
     utils.assign(journal_config.data, {
-        frame = self.frame
+        frame = self.frame,
+        toc = {
+            width = toc.frame.w,
+            visible = toc.visible
+        }
     })
     journal_config:write()
 end
 
+function JournalWindow:loadConfig()
+    if not self.save_layout then
+        return nil, false, 25
+    end
+
+    local window_frame = copyall(journal_config.data.frame or {})
+    local table_of_contents = copyall(journal_config.data.toc or {
+        width=20,
+        visible=false
+    })
+
+    return window_frame, table_of_contents.visible or false, table_of_contents.width or 25
+end
+
+function JournalWindow:onPanelResizeBegin()
+    self.resizing_panels = true
+end
+
+function JournalWindow:onPanelResizeEnd()
+    self.resizing_panels = false
+    self:ensurePanelsRelSize()
+
+    self:updateLayout()
+end
+
+function JournalWindow:onRenderBody(painter)
+    if self.resizing_panels then
+        self:ensurePanelsRelSize()
+        self:updateLayout()
+    end
+
+    return JournalWindow.super.onRenderBody(self, painter)
+end
+
+function JournalWindow:ensurePanelsRelSize()
+    local toc_panel = self.subviews.table_of_contents_panel
+    local editor = self.subviews.journal_editor
+    local divider = self.subviews.table_of_contents_divider
+
+    toc_panel.frame.w = math.min(
+        math.max(toc_panel.frame.w, toc_panel.resize_min.w),
+        self.frame.w - editor.resize_min.w
```

Comment:
```sh
/home/myk/src/dfhack/scripts/gui/journal.lua:211: attempt to perform arithmetic on a nil value (field 'w')
stack traceback:
        /home/myk/src/dfhack/scripts/gui/journal.lua:211: in method 'ensurePanelsRelSize'
        /home/myk/src/dfhack/scripts/gui/journal.lua:218: in local 'fun'
        ./hack/lua/class.lua:110: in function <./hack/lua/class.lua:105>
        (...tail calls...)
        ./hack/lua/gui.lua:790: in method 'updateLayout'
        ./hack/lua/gui.lua:779: in method 'updateSubviewLayout'
        ./hack/lua/gui.lua:799: in method 'updateLayout'
        ./hack/lua/gui.lua:976: in method 'onResize'
        ./hack/lua/gui.lua:902: in local 'fun'
```

Note that `frame.w` is not only not guaranteed to exist, but also does not represent the current state of the window. The frame represents the *desired* dimensions, not the *actual* dimensions. I think you might want `self.frame_body.width` or `self.frame_parent_rect.width`.

both fields are set as part of `updateLayout`, so they might not exist on the first call to `preUpdateLayout`.

Reply:
Can you try to repeat the issue and explain when you exactly get this error?

There was a refactor error in the code for 20 minutes, you could test it on the bugged version.

Reply:
with my understand as I always set `journal_editor` and `table_of_contents_panel` in `JournalWindow:init` - they should be available from the beginning in my case.

Reply:
I see now, its on server automatic tests. Checking.

Reply:
should be fixed now - it was a problem with missing defaults for layout restoring.